config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java

```diff
@@ -32,59 +41,147 @@
+                        bundleEntities, projectGroupName, projectVersion);
+
+                return ImmutablePair.of(annotatedBundle,
+                        bundleMetadata);
+            }
+            return null;
+        } else {
+            return ImmutablePair.of(bundleDocumentBuilder.build(document,
+                    entities), null);
+        }
+    }
+
+    private List<Entity> renameNonReusableEntities(List<Entity> entityList, Bundle bundle, AnnotatedEntity annotatedEntity, String projectGroupName, String projectVersion) {
+        List<Entity> bundleEntities = new ArrayList<>();
+        final Map<String, Policy> policyMap = bundle.getPolicies();
+        final Map<String, Encass> encassMap = bundle.getEncasses();
+        for(Entity entity : entityList){
+            if(EntityTypes.ENCAPSULATED_ASSERTION_TYPE.equals(entity.getType()) || EntityTypes.POLICY_TYPE.equals(entity.getType())) {
+                Encass encassEntity = encassMap.get(entity.getName());
+                Policy policyEntity = policyMap.get(entity.getName());
+                Set<Annotation> annotations = policyEntity != null ? policyEntity.getAnnotations() : encassEntity.getAnnotations();
+                if (annotations != null && annotations.stream().anyMatch(annotation -> ANNOTATION_TYPE_REUSABLE_ENTITY.equals(annotation.getType()))) {
+                    bundleEntities.add(entity);
+                } else {
+                    bundleEntities.add(getUniqueEntity(entity, annotatedEntity, encassMap, projectGroupName, projectVersion));
+                }
 
-        Map<String, Pair<Element, BundleMetadata>> artifacts = buildAnnotatedEntities(entities, bundle, document, projectName,
-                projectGroupName, projectVersion);
-        if (artifacts.isEmpty()) {
-            artifacts.put(StringUtils.isBlank(projectVersion) ? projectName : projectName + "-" + projectVersion,
-                    ImmutablePair.of(bundleDocumentBuilder.build(document, entities), null));
+            } else {
+                bundleEntities.add(entity);
+            }
         }
+        return bundleEntities;
+    }
 
-        return artifacts;
+    private Entity getUniqueEntity(final Entity entity, final AnnotatedEntity annotatedEntity, final Map<String, Encass> encassMap, final String projectName, final String projectVersion) {
+        final String nameWithPath = entity.getName();
+        final String uniqueName = getUniqueName(projectName, projectVersion, annotatedEntity, nameWithPath);
+        final String uniqueNameWithPath = PathUtils.extractPath(nameWithPath) + uniqueName;
+        Element entityXml = null;
+        if(EntityTypes.POLICY_TYPE.equals(entity.getType())){
+            entityXml = getUniquePolicyXml(entity.getXml(), annotatedEntity, encassMap, projectName, projectVersion, uniqueName);
+        } else if(EntityTypes.ENCAPSULATED_ASSERTION_TYPE.equals(entity.getType())){
+            entityXml = (Element) entity.getXml().cloneNode(true);
```

Comment:
entity.getXml gives clone Element every time when you call

Reply:
done